### PR TITLE
Minimize the calls in useSelect by subscribing to only the stores needed

### PR DESCRIPTION
**Pull request by kevin940726**
## Description

Minimize the calls of the selectors in `useSelect` by only subscribing to the stores needed.

### The problem

Currently, `useSelect` works by subscribing to the `registry` instance and calls the `mapSelect` (the first argument of `useSelect`) function every time the store changes. It works fine, but it also fires many unnecessary `mapSelect` calls when we know that the store updates won't cause a re-render.

Most components subscribe to the default `registry`, and most stores also registered to that same `registry`. Let's say that some component has a `counter` store and multiple other stores like `core`, `core/editor`, `core/block-editor`, etc. The `counter` store has a selector `getCounter`, which simply gets the `count` state from the store. A component using this selector could look like this.

```js
function TestComponent() {
	const { count } = useSelect( select => ({
		count: select( 'counter' ).getCounter(),
	}), [] );

	return count;
}
```

It's clear that this component doesn't care about state updates from stores other than `counter`. However, the selector still gets called when we dispatch any actions no matter if the action doesn't belong to the `counter` store.

Fortunately, since the `count` state from `getCounter()` doesn't change, `<TestComponent>` doesn't have to re-render. Still, it's an unnecessary call of the `getCounter` selector. In this specific case, it doesn't matter much, but it would matter if the `mapSelect` function is very complex or slow. Especially when we have `resolver` for each `selector`, which could potentially cause unwanted side-effects to be introduced.

### The solution

This PR propose a solution to minimize the calls of the `mapSelect` function as much as possible. The idea is to only call the `mapSelect` function if and only if the store this `mapSelect` has referenced to gets updated.

For instance,

```js
useSelect( select => ({
	count: select( 'counter' ).getCounter(),
}), [] );
```

this `useSelect` only cares about updates within the `counter` store itself and nothing else. We only have to re-run the `mapSelect` function whenever the state from the `counter` store updates.

Another example,

```js
useSelect( select => ({
	count: select( 'counter' ).getCounter(),
	isPublished: select( 'core/editor' ).isCurrentPostPublished(),
}), [] );
```

this `useSelect` cares about state updates from two different stores: `counter` and `core/editor`. Whenever either one of the store updates, we need to re-run the `mapSelect` function.

This is done by _trapping_ the `select` function of the `registry` to keep track of every store it has called. It works very similarly like monkey-patching the `select` function as follow (the actual implementation is slightly different because of some issues with legacy features with `use` and `withPlugins`):

```js
const listeningStores = [];

const originalSelect = registry.select;
registry.select = function trapSelect( storeKey ) {
	listeningStores.push( storeKey );
	return originalSelect.call( this, storeKey );
};
```

Whenever `reigstry.select( storeKey )` is called, we'll push the `storeKey` to the `listeningStores`. Later, we can then only subscribe to those stores rather than subscribing to the whole `registry`.

```diff js
- registry.subscribe( onChange );
+ listeningStores.forEach( storeKey => registry.stores[ storeKey ].subscribe( onChange ) );
```

## How has this been tested?

Added a bunch of unit tests and also fixed some tests.
End-to-end tests should also be passing.

### Performance tests

```
>> post-editor

┌─────────┬───────────┬────────────┬────────────┬─────────────┬─────────────┬────────────┬─────────────┐
│ (index) │   load    │    type    │  minType   │   maxType   │    focus    │  minFocus  │  maxFocus   │
├─────────┼───────────┼────────────┼────────────┼─────────────┼─────────────┼────────────┼─────────────┤
│ This PR │ '6733 ms' │ '44.5 ms'  │ '40.78 ms' │ '100.23 ms' │ '98.93 ms'  │ '73.53 ms' │ '131.48 ms' │
│ master  │ '7360 ms' │ '53.59 ms' │ '46.28 ms' │ '113.3 ms'  │ '105.19 ms' │ '65.89 ms' │ '143.31 ms' │
└─────────┴───────────┴────────────┴────────────┴─────────────┴─────────────┴────────────┴─────────────┘

>> site-editor

┌─────────┬──────────────┐
│ (index) │     load     │
├─────────┼──────────────┤
│ This PR │  '1243 ms'   │
│ master  │ '1255.33 ms' │
└─────────┴──────────────┘
```

As seen in the report, this PR out-performs every test. However, the difference is not that big though, it might be just a micro-optimization in the end.

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by adamziel** [comment]
@kevin940726 it doesn't account for selectors like

```js
export const getWidgets = createRegistrySelector( ( select ) => () => {
	const widgets = select( 'core' ).getEntityRecords(
		'root',
		'widget',
		buildWidgetsQuery()
	);

	return keyBy( widgets, 'id' );
} );
```

To make it work, you'd need to somehow hook into actual `select( store )`, store all the synchronous calls, and access them in `useSelect`. I wonder it that could get somehow mixed up with an unrelated `mapSelect` too.

**Comment by adamziel** [comment]
@kevin940726 I played with it more and came up with this: https://github.com/WordPress/gutenberg/pull/26733 - still not working properly though

**Comment by youknowriad** [comment]
Does it support this as well https://github.com/WordPress/gutenberg/pull/26724#issuecomment-722330948 ?

**Comment by kevin940726** [comment]
@youknowriad Yep, it's also in the unit tests :).

**Comment by adamziel** [comment]
@kevin940726 I think this PR is obsolete now that https://github.com/WordPress/gutenberg/pull/26866 was merged

**Comment by kevin940726** [comment]
> @kevin940726 I think this PR is obsolete now that #26866 was merged

Thx for the notice! Note to myself that I'll try to run some of the unit tests introduced here after rebasing and close this 👍.

**Comment by youknowriad** [comment]
Actually, I'm considering reverting the `@wordpress/stan` temporarily (to give me time to do more performance tests as I discovered some issues that need clarification).  Maybe we can land this smaller approach after the revert.

**Comment by kevin940726** [comment]
@youknowriad Thx for the inputs! I've rebased the branch and updated the performance report in the description.

File: packages/data/src/registry.js [hunk 1]
```diff
@@ -214,6 +223,7 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
 	}
 
 	let registry = {
+		parent,
```
Comment (youknowriad) [inline packages/data/src/registry.js:226]: Why do we need parent in the API of the registry?

Comment (kevin940726) [reply packages/data/src/registry.js:226]: To be able to traverse up the registries tree to find the correct registry which registered the store.
See https://github.com/WordPress/gutenberg/pull/26724/files#diff-842c1e34fc3a1df56dc553d99766a9f86a3e0e141a99632471ed462e2b04035fR39-R54

Comment (kevin940726) [reply packages/data/src/registry.js:226]: I wonder if there's a better alternative? I dislike that we have to add additional API to `registry`, but it's the easiest way I can think of for now.

Comment (youknowriad) [reply packages/data/src/registry.js:226]: In my mind even `stores` should be private (though it's not new).;

I think what you really want is a function to subscribe to just a given store so maybe `registry.__experimentalSubscribeStore( storeName, handler )` 

or potentially  update `registry.subscribe` to allow  `registry.subscribe( storeName, handler )`

Comment (kevin940726) [reply packages/data/src/registry.js]: Cool, cool! I pushed 9295f6d to use a `WeakMap` instead to store these values privately (without breaking changes of course).

Comment (kevin940726) [reply packages/data/src/registry.js]: The solution isn't perfect, have to do some workarounds here and there due to the legacy `withPlugins` and `use` methods.

An alternative solution is to use an `_id` and store it in the registry object, the downside obviously is that we're now introducing a private API to the registry object though.

File: packages/data/src/registry.js [hunk 2]
```diff
@@ -279,5 +321,16 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
 		parent.subscribe( globalListener );
 	}
 
-	return withPlugins( registry );
+	const storesMap = {
+		stores,
+		parent,
+	};
+	StoresMap.set( registry, storesMap );
```
Comment (youknowriad) [inline packages/data/src/registry.js:333]: Do we really need this map, I mean can't we just have `__experimentalSubscribeStores` call the `__experimentalSubscribeStores` for  the parent registries when the store is not available on the current registry?

Comment (kevin940726) [reply packages/data/src/registry.js:333]: Shoot! We can! I have no idea why I haven't thought about this way simpler solution 😅, thx!

Comment (kevin940726) [reply packages/data/src/registry.js]: Pushed ecc8d74.

**Comment by youknowriad** [comment]
It seems the performance impact here is good but not as good as what we had in the atom's approach. 
I believe the difference with atoms, is that there was no need for the selectors to run again inside `useEffect` on first mount. This optimization though, doesn't seem to be possible here.

**Review by youknowriad — APPROVED** [review]
LGTM

**Review by adamziel — APPROVED** [review]
I like the performance gains, and the code looks good as well. 🚢

**Comment by adamziel** [comment]
> I believe the difference with atoms, is that there was no need for the selectors to run again inside useEffect on first mount. This optimization though, doesn't seem to be possible here.

Our stores are large so it's normal to have selectors "watching" a branch that stays the same, that's another difference with the atoms approach - only relevant changes bubble up to their observers.

**Pull request merged**